packages/core/src/derived/processor.ts

```diff
@@ -0,0 +1,237 @@
+const EMPTY_MAPPER = '' as Resource<DocumentMapper>
+
+type Descr = DerivedDataDescriptor<Doc, DerivedData>
+interface CachedDoc {
+  doc?: Doc
+  resolve: () => Promise<Doc>
+}
+interface ObjTx {
+  objectId: Ref<Doc>
+  objectClass: Ref<Class<Doc>>
+}
+/**
+ * Register derived data mapper globally.
+ */
+export function registerMapper (id: Resource<DocumentMapper>, mapper: DocumentMapper): void {
+  derivedDataMappers.set(id, mapper)
+}
+
+/**
+/**
+ * Allow to generate derived data with rules and mappers.
+ */
+export class DerivedDataProcessor extends TxProcessor {
+  descrs: DescriptorMap
+
+  private constructor (readonly model: ModelDb, readonly hierarchy: Hierarchy, readonly storage: Storage) {
+    super()
+    this.descrs = new DescriptorMap(hierarchy)
+  }
+
+  /**
+   * Obtain initial set of descriptors and have derived data up to date.
+   */
+  static async create (model: ModelDb, hierarchy: Hierarchy, storage: Storage): Promise<DerivedDataProcessor> {
+    const processor = new DerivedDataProcessor(model, hierarchy, storage)
+    const descriptors = await model.findAll(core.class.DerivedDataDescriptor, {})
+
+    for (const d of descriptors) {
+      processor.descrs.update(d)
+    }
+    return processor
+  }
+
+  protected async txCreateDoc (tx: TxCreateDoc<Doc>): Promise<void> {
+    if (this.hierarchy.isDerived(tx.objectClass, core.class.DerivedDataDescriptor)) {
+      await this.updateDescriptor(tx.objectId as Ref<Descr>, tx.modifiedBy)
+      return
+    }
+
+    // Obtain descriptors an traverse build derived data
+    const descriptors = this.descrs.getByClass(tx.objectClass)
+    const ops = withOperations(tx.modifiedBy, this.storage)
+    const doc: CachedDoc = {
+      resolve: async () => await Promise.resolve(TxProcessor.createDoc2Doc(tx))
+    }
+
+    for (const d of descriptors) {
+      let results = (await this.applyMapper(d, tx)) ?? (await this.applyRule(d, tx, doc))
+      results = this.withDescrId(results, d._id, tx.objectId, tx.objectClass)
+      await this.applyDerivedData([], results, ops)
+    }
+  }
+
+  private async updateDescriptor (objectId: Ref<Descr>, modifiedBy: Ref<Account>): Promise<void> {
+    this.descrs.remove(objectId)
+    // New descriptor construction
+    const descr = this.model.getObject(objectId)
+    this.descrs.update(descr)
+
+    await this.refreshDerivedData(descr, modifiedBy)
+  }
+
+  protected async txUpdateDoc (tx: TxUpdateDoc<Doc>): Promise<void> {
+    if (this.hierarchy.isDerived(tx.objectClass, core.class.DerivedDataDescriptor)) {
+      await this.updateDescriptor(tx.objectId as Ref<Descr>, tx.modifiedBy)
+      return
+    }
+
+    const descriptors = this.descrs.getByClass(tx.objectClass)
+    const ops = withOperations(tx.modifiedBy, this.storage)
+
+    // We expect storage is alrady updated before derived data is being processed.
+    const doc: CachedDoc = {
+      resolve: async () => (await this.storage.findAll(tx.objectClass, { _id: tx.objectId }))[0]
```

Comment:
Right now I've decided to take source object in case of it update for processing.
Potentially we could check if rules has effect on change and only in this case take object from underline storage.